<h1 id="tutorial-for-vcfhunter">Tutorial for VcfHunter</h1>
<p>This tutorial go through all steps from variant calling from DNAseq to marker generation for genetic mapping. <strong><em>VcfHunter tools</em></strong></p>
<p>Go to the VcfHunter folder (Scripts can be run from any folder but the command lines in this tutorial assume you are in this folder and that you have python 2.7 and 3 versions).</p>
<h3 id="available-data">Available data:</h3>
<p><em>data/reference/</em> is a folder containing a fasta file with the reference sequence and a gff file containing the reference sequence annotation <em>data/reads/</em> is a folder containing DNAseq paired reads simulated from 3 gene pools of diploid ancestral accessions (11, 10 and 9 accessions from each gene pool respectively). <em>data/config/</em> is a folder containing one configuration files needed for the DNAseq variant calling.</p>
<p>A specific folder as been created to test the tools (<em>TestTools</em>). Go to this folder to run all command lines of this toolbox. But these tools can also be launched from everywhere you have permission to write, you will only have to change path to the configuration files.</p>
<p>For the following example, we will assume that you have a computer with 8 processors. If it is not the case or is you want to use less processors than available you should/can change the number of processors you allow the program to use.</p>
<p>In this tutorial we assume that you have performed the preliminary steps of checking your reads and filtered if necessary. After that, it is time to begin the variant calling.</p>
<h2 id="a---variant-calling">A - Variant calling</h2>
<p>Because different mapping tools are needed to perfom the mapping depending on the type of data (DNA), two different pipeline shoold be use for read mapping and post mapping processing.</p>
<h3 id="variant-calling">Variant calling</h3>
<p>a - Mapping of the DNA reads: In this step, reads are aligned against the reference sequence using BWA mem algorithm</p>
<pre><code>python2 ../bin/process_reseq_1.0.py -c ../data/config/DNAseq.conf -t 8 -p DNAseq -s a</code></pre>
<p>40 folders have been created containing each:</p>
<ul>
<li>*.sam.stat files with mapping statistics</li>
<li>*_merged.bam file containing aligned read merged in one bam file</li>
<li>*_merged.bai file (index of the bam file)</li>
<li>STAT folder containing a .html file summarising statistics on libraries.</li>
</ul>
<p>b - Removing duplicates reads: In this step, read duplicates are removed</p>
<pre><code>python2 ../bin/process_reseq_1.0.py -c ../data/config/DNAseq.conf -t 8 -p DNAseq -s b</code></pre>
<p>3 new files have been added in each folders:</p>
<ul>
<li>*_rmdup.bam file containing non redundant reads resulting from PCR and optical duplicates</li>
<li>*_rmdup.bai file (index of the bam file)</li>
<li>*_duplicate file summarising read duplication statistics</li>
</ul>
<p>c - Removing duplicates reads: In this step, read are realigned around indels</p>
<pre><code>python2 ../bin/process_reseq_1.0.py -c ../data/config/DNAseq.conf -t 8 -p DNAseq -s c</code></pre>
<p>2 new files have been added in each folders:</p>
<ul>
<li>*_realigned.bam file reads realigned around indels</li>
<li>*_realigned.bai file (index of the bam file)</li>
</ul>
<p>d - Base recalibration: In this step, reads base sequencing quality are recalculated. <strong>This step is recommended by GATK best practice but we do not recommend to use it if you use our pipeline.</strong></p>
<pre><code>python2 ../bin/process_reseq_1.0.py -c ../data/config/DNAseq.conf -t 8 -p DNAseq -s d</code></pre>
<p>5 new files have been added in each folders:</p>
<ul>
<li>*_real_recal.bam file reads bases recalibrated</li>
<li>*_real_recal.bai file (index of the bam file)</li>
<li>*_recalibration_plot.pdf file</li>
<li>*_selected.vcf file containing variant sites used for base recalibration (if no vcf have been provided)</li>
<li>*_selected.vcf.idx file (index of the vcf file)</li>
</ul>
<p>e - Allele counting: In this step, reads are used to count for each covered sites the number of reads supporting each bases (A,T,G,C,N,*=deletion).</p>
<pre><code>python2 ../bin/process_reseq_1.0.py -c ../data/config/DNAseq.conf -t 8 -p DNAseq -s e</code></pre>
<p>Several new files have been added in each folders:</p>
<ul>
<li>*.gz files recording, for each chromosomes/sequences in the fasta provided as reference, the count of the number of reads supporting each bases (A,T,G,C,N,*=deletion) at each covered site.</li>
</ul>
<p>f - Vcf generation: generate the vcf for all accessions in the configuration file</p>
<pre><code>python2 ../bin/process_reseq_1.0.py -c ../data/config/DNAseq.conf -t 8 -p DNAseq -s f</code></pre>
<p>Several new files in the current directory:</p>
<ul>
<li>prefix_*_all_allele_count.vcf file for each chromosomes/sequences in the fasta provided as reference.</li>
</ul>
<p>g - Vcf merging: generate a single vcf from all chromosome/sequences in the fasta provided as reference. This step is not needed here has only one sequence is passed but you can try the command line anyway ;-)</p>
<pre><code>python2 ../bin/process_reseq_1.0.py -c ../data/config/DNAseq.conf -t 8 -p DNAseq -s g</code></pre>
<p>1 new file is created in the current directory:</p>
<ul>
<li>prefix_all_allele_count.vcf file</li>
</ul>
<p>h - Statistics: Compute mapping statistics</p>
<pre><code>python2 ../bin/process_reseq_1.0.py -c ../data/config/DNAseq.conf -t 8 -p DNAseq -s h</code></pre>
<p>Several new files in the current directory:</p>
<ul>
<li>*_lib.stats file containing mapping statistics per libraries</li>
<li>*_acc.stats file containing mapping statistics per accessions</li>
</ul>
<p><strong><em>All the steps can be lauched in one command line. First remove the generated files:</em></strong></p>
<pre><code>rm -rf *</code></pre>
<p>Run the pipeline again (without step d)</p>
<pre><code>python2 ../bin/process_reseq_1.0.py -c ../data/config/DNAseq.conf -t 8 -p DNAseq -s abcefgh</code></pre>
<h2 id="b---vcf-prefiltering">B - VCF prefiltering</h2>
<p>Once the variant calling is performed, it is now time to filter the calling. This is a necessary step with this pipeline because all variant sites are reported (one read, on one accession supporting a variant). The phylosophie of <strong><em>process_RNAseq</em></strong> and <strong><em>process_reseq</em></strong> is to report all variant sites and then decide to filter the vcf (what is error and what is not) based on understandable and simple parametres using <strong><em>VcfPreFilter</em></strong> tool. For the following step we will work on the vcf file generated on both DNA and RNA seq data. VcfPreFilter can be launched as followed:</p>
<pre><code>python2 ../bin/VcfPreFilter.1.0.py -v DNAseq_all_allele_count.vcf -m 10 -M 10000 -f 0.05 -c 3 -o DNAseq_prefiltered.vcf</code></pre>
<p>The outpout is a vcf file (named as filled in -o option) in which the variant line are filtered as followed:</p>
<ul>
<li>1 - an accession of a variant line is considered only if its site coverage is comprised in [-m : -M] (in this example [10:10000]),</li>
<li>2 - for each considered accession, an allele is considered only if its frequency at in the accession if equal or greater than -f (in this example: 0.05) and its coverage is equal or greater than -c (in this example: 3),</li>
<li>3 - the line is kept only if their is at least one allele different from the reference kept in at least one accession at the end of the process,</li>
</ul>
<p>The variant calling is recalculated <strong>for each accessions</strong> based on selected alleles (even accessions which are not enough covered according our parameters will have a genotype). An additional tag (GC) is added in the FORMAT column of the VCF. This tag is calculated as followed : -log10(best genotype probability) + log10(second best genotype probability) and give an idea of the quality of the calling at the datapoint.</p>
